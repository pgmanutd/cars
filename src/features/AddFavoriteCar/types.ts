import { Car } from 'features/CarInfo/types';

export interface FavoriteCars {
  [stockNumber: string]: Car;
}
