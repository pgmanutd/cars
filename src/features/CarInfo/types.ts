export type StockNumberParam = string;

export interface Car {
  stockNumber: number;
  manufacturerName: string;
  modelName: string;
  mileage: {
    number: number;
    unit: string;
  };
  fuelType: string;
  color: string;
  pictureUrl: string;
}

export interface CarResponse {
  car: Car;
}

export interface FavoriteCars {
  [stockNumber: string]: Car;
}
