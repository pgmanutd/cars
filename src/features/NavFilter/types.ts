export type ColorQuery = string;
export type ManufacturerQuery = string;

export type Color = string;
export type Colors = Color[];

export interface ColorsResponse {
  colors: Colors;
}

export interface Manufacturer {
  name: string;
  models: {
    name: string;
  }[];
}
export type Manufacturers = Manufacturer[];

export interface ManufacturersResponse {
  manufacturers: Manufacturers;
}
