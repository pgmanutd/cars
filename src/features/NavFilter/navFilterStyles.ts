import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      border: `1px solid ${theme.palette.divider}`,
      padding: theme.spacing(3),
    },
    formControl: {
      width: '100%',
      marginBottom: theme.spacing(3),
    },
    selectEmpty: {
      marginTop: theme.spacing(3),
    },
    filterButtonContainer: {
      display: 'flex',
      justifyContent: 'flex-end',
    },
  }),
);
