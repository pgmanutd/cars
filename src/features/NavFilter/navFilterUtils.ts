import { QUERY_VALUES } from 'constants/appConstants';

import routePaths from 'constants/routePaths';

import { ColorQuery, ManufacturerQuery } from './types';

export const getFilterButtonHref = ({
  selectedColor,
  selectedManufacturer,
}: {
  selectedColor: ColorQuery;
  selectedManufacturer: ManufacturerQuery;
}) => {
  const query = {
    [QUERY_VALUES.colorFilter]: selectedColor,
    [QUERY_VALUES.manufacturerFilter]: selectedManufacturer,
  };

  return `${routePaths.cars}?${new URLSearchParams(
    query as Record<string, string>,
  )}`;
};

export const pickTargetValue = (event: React.ChangeEvent<{ value: unknown }>) =>
  event.target.value;
