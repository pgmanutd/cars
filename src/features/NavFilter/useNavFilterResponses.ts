import apiPaths from 'constants/apiPaths';

import useFetch from 'hooks/useFetch';

import {
  ColorsResponse,
  ManufacturersResponse,
  Colors,
  Manufacturers,
} from './types';

const useNavFilterResponses = () => {
  const [colorsResponse] = useFetch<ColorsResponse>({
    initialUrl: apiPaths.colors(),
  });

  // NOTE: Not handling error case as of now. Showing skeleton instead in case of error.
  const areColorsResponseLoading =
    colorsResponse.isLoading || !colorsResponse.data || colorsResponse.error;
  const colorsResponseValue = colorsResponse.data?.colors ?? ([] as Colors);

  const [manufacturersResponse] = useFetch<ManufacturersResponse>({
    initialUrl: apiPaths.manufacturers(),
  });

  const areManufacturersResponseLoading =
    manufacturersResponse.isLoading ||
    !manufacturersResponse.data ||
    manufacturersResponse.error;
  const manufacturersResponseValue =
    manufacturersResponse.data?.manufacturers ?? ([] as Manufacturers);

  const areNavFiltersResponseLoading = !!(
    areColorsResponseLoading || areManufacturersResponseLoading
  );

  return {
    areColorsResponseLoading,
    colorsResponseValue,
    areManufacturersResponseLoading,
    manufacturersResponseValue,
    areNavFiltersResponseLoading,
  };
};

export default useNavFilterResponses;
