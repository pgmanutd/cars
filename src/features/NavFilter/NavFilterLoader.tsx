import React from 'react';
import Box from '@material-ui/core/Box';
import Skeleton from '@material-ui/lab/Skeleton';

import { useStyles } from './navFilterStyles';

export interface NavFilterLoaderProps {
  areColorsResponseLoading: boolean;
  areManufacturersResponseLoading: boolean;
  areNavFiltersResponseLoading: boolean;
}

const NavFilterLoader: React.FC<NavFilterLoaderProps> = ({
  areColorsResponseLoading,
  areManufacturersResponseLoading,
  areNavFiltersResponseLoading,
}) => {
  const classes = useStyles();

  return (
    <Box
      data-testid="NavFilterLoader"
      component="aside"
      className={classes.root}
    >
      {areColorsResponseLoading && (
        <Skeleton data-testid="NavFilterColorSkeleton" height={48} />
      )}
      {areManufacturersResponseLoading && (
        <Skeleton data-testid="NavFilterManufacturerSkeleton" height={48} />
      )}
      <Box className={classes.filterButtonContainer}>
        {areNavFiltersResponseLoading && (
          <Skeleton
            data-testid="NavFilterButtonSkeleton"
            height={70}
            width={92}
          />
        )}
      </Box>
    </Box>
  );
};

export default NavFilterLoader;
