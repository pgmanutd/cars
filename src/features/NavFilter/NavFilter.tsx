import React, { memo, useState, useCallback } from 'react';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import FormControl from '@material-ui/core/FormControl';
import Select from '@material-ui/core/Select';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Skeleton from '@material-ui/lab/Skeleton';
import { Link as RouterLink } from 'react-router-dom';
import _compose from 'lodash/fp/compose';

import { useTranslate } from 'features/Translate';

import { ColorQuery, ManufacturerQuery } from './types';
import { getFilterButtonHref, pickTargetValue } from './navFilterUtils';
import useNavFilterResponses from './useNavFilterResponses';
import { useStyles } from './navFilterStyles';

export interface NavFilterProps {
  selectedColor: ColorQuery;
  selectedManufacturer: ManufacturerQuery;
}

const NavFilter: React.FC<NavFilterProps> = ({
  selectedColor,
  selectedManufacturer,
  ...restProps
}) => {
  const classes = useStyles();
  const { translate } = useTranslate();

  const [color, setColor] = useState(selectedColor);
  const [manufacturer, setManufacturer] = useState(selectedManufacturer);

  const handleColorChange = useCallback(
    _compose(setColor, pickTargetValue),
    [],
  );

  const handleManufacturerChange = useCallback(
    _compose(setManufacturer, pickTargetValue),
    [],
  );

  const {
    areColorsResponseLoading,
    colorsResponseValue,
    areManufacturersResponseLoading,
    manufacturersResponseValue,
    areNavFiltersResponseLoading,
  } = useNavFilterResponses();

  return (
    <Box
      data-testid="NavFilter"
      {...restProps}
      component="section"
      className={classes.root}
    >
      {areColorsResponseLoading ? (
        <Skeleton data-testid="NavFilterColorSkeleton" height={48} />
      ) : (
        <FormControl
          data-testid="NavFilterColor"
          className={classes.formControl}
        >
          <InputLabel shrink id="NavFilterColorLabel">
            {translate('features.NavFilter.colorLabel')}
          </InputLabel>
          <Select
            labelId="NavFilterColorLabel"
            id="NavFilterColorLabelSelect"
            value={color}
            onChange={handleColorChange}
            displayEmpty
            className={classes.selectEmpty}
          >
            <MenuItem value="">
              <em>{translate('features.NavFilter.defaultColorText')}</em>
            </MenuItem>
            {colorsResponseValue.map(colorValue => (
              <MenuItem key={colorValue} value={colorValue}>
                {colorValue}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      )}
      {areManufacturersResponseLoading ? (
        <Skeleton data-testid="NavFilterManufacturerSkeleton" height={48} />
      ) : (
        <FormControl
          data-testid="NavFilterColorManufacturer"
          className={classes.formControl}
        >
          <InputLabel shrink id="NavFilterManufacturerLabel">
            {translate('features.NavFilter.manufacturerLabel')}
          </InputLabel>
          <Select
            labelId="NavFilterManufacturerLabel"
            id="NavFilterManufacturerLabelSelect"
            value={manufacturer}
            onChange={handleManufacturerChange}
            displayEmpty
            className={classes.selectEmpty}
          >
            <MenuItem value="">
              <em>{translate('features.NavFilter.defaultManufacturerText')}</em>
            </MenuItem>
            {manufacturersResponseValue.map(({ name: manufacturerValue }) => (
              <MenuItem key={manufacturerValue} value={manufacturerValue}>
                {manufacturerValue}
              </MenuItem>
            ))}
          </Select>
        </FormControl>
      )}
      <Box className={classes.filterButtonContainer}>
        {areNavFiltersResponseLoading ? (
          <Skeleton
            data-testid="NavFilterButtonSkeleton"
            height={70}
            width={92}
          />
        ) : (
          <Button
            data-testid="NavFilterButton"
            component={RouterLink}
            variant="contained"
            color="primary"
            size="large"
            to={`${getFilterButtonHref({
              selectedColor: color,
              selectedManufacturer: manufacturer,
            })}`}
          >
            {translate('features.NavFilter.filterButtonText')}
          </Button>
        )}
      </Box>
    </Box>
  );
};

export default memo(NavFilter);
