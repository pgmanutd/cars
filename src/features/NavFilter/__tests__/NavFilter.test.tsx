import React from 'react';
import { waitForElement } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route } from 'react-router-dom';

import apiPaths from 'constants/apiPaths';
import routePaths from 'constants/routePaths';
import { QUERY_VALUES } from 'constants/appConstants';

import { renderWithProviders } from 'utils/testUtils';

import { ColorQuery, ManufacturerQuery } from '../types';
import NavFilter from '../NavFilter';

describe('<NavFilter />', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  const setup = ({
    selectedColor,
    selectedManufacturer,
  }: {
    selectedColor: ColorQuery;
    selectedManufacturer: ManufacturerQuery;
  }) => {
    const initialRoute = routePaths.cars;

    const { renderResult, history } = renderWithProviders(
      <Route path={initialRoute}>
        <NavFilter
          selectedColor={selectedColor}
          selectedManufacturer={selectedManufacturer}
        />
      </Route>,
      {
        routerConfig: {
          route: initialRoute,
        },
      },
    );

    return {
      renderResult,
      history,
    };
  };

  const setupForSuccessResponse = async () => {
    const selectedColor = '';
    const successColorsResponse = {
      colors: ['white'],
    };

    const selectedManufacturer = '';
    const successManufacturerResponse = {
      manufacturers: [
        {
          name: 'Audi',
          models: [
            {
              name: 'ABCD',
            },
          ],
        },
      ],
    };

    fetchMock
      .doMockOnceIf(apiPaths.colors())
      .mockResponseOnce(JSON.stringify(successColorsResponse));

    fetchMock
      .doMockOnceIf(apiPaths.manufacturers())
      .mockResponseOnce(JSON.stringify(successManufacturerResponse));

    const { renderResult, history } = setup({
      selectedColor,
      selectedManufacturer,
    });

    await waitForElement(() => renderResult.queryByTestId('NavFilterButton'));

    return {
      renderResult,
      history,
      successColorsResponse,
      successManufacturerResponse,
    };
  };

  it('should render the component and matches it against stored snapshot for success response', async () => {
    const { renderResult } = await setupForSuccessResponse();

    expect(renderResult.asFragment()).toMatchSnapshot();
  });

  it('should change location path to selected values on "Filter" button press', async () => {
    const {
      renderResult,
      history,
      successColorsResponse,
      successManufacturerResponse,
    } = await setupForSuccessResponse();

    const color = successColorsResponse.colors[0];
    const manufacturer = successManufacturerResponse.manufacturers[0].name;

    userEvent.click(renderResult.getByText('All car colors'));
    userEvent.click(renderResult.getByText(color));
    userEvent.click(renderResult.getByText('All manufacturers'));
    userEvent.click(renderResult.getByText(manufacturer));

    userEvent.click(renderResult.getByText('Filter'));

    const searchParams = new URLSearchParams(history.location.search);

    expect([
      searchParams.get(QUERY_VALUES.colorFilter),
      searchParams.get(QUERY_VALUES.manufacturerFilter),
    ]).toStrictEqual([color, manufacturer]);
  });

  it('should render the component and matches it against stored snapshot for failed response', async () => {
    const selectedColor = 'white';
    const selectedManufacturer = 'Audi';
    const errorResponse = new Error('Some error');

    fetchMock.doMockOnceIf(apiPaths.colors()).mockRejectOnce(errorResponse);
    fetchMock
      .doMockOnceIf(apiPaths.manufacturers())
      .mockRejectOnce(errorResponse);

    const { renderResult } = setup({
      selectedColor,
      selectedManufacturer,
    });

    await waitForElement(() =>
      renderResult.queryByTestId('NavFilterButtonSkeleton'),
    );

    expect(renderResult.asFragment()).toMatchSnapshot();
  });
});
