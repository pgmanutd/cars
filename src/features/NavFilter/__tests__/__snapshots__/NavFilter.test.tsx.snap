// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavFilter /> should render the component and matches it against stored snapshot for failed response 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-325 makeStyles-root-320"
    data-testid="NavFilter"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
      data-testid="NavFilterColorSkeleton"
      style="height: 48px;"
    />
    <span
      class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
      data-testid="NavFilterManufacturerSkeleton"
      style="height: 48px;"
    />
    <div
      class="MuiBox-root MuiBox-root-334 makeStyles-filterButtonContainer-323"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
        data-testid="NavFilterButtonSkeleton"
        style="width: 92px; height: 70px;"
      />
    </div>
  </section>
</DocumentFragment>
`;

exports[`<NavFilter /> should render the component and matches it against stored snapshot for success response 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-6 makeStyles-root-1"
    data-testid="NavFilter"
  >
    <div
      class="MuiFormControl-root makeStyles-formControl-2"
      data-testid="NavFilterColor"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
        data-shrink="true"
        id="NavFilterColorLabel"
      >
        Color
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectEmpty-3 MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="NavFilterColorLabel NavFilterColorLabelSelect"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="NavFilterColorLabelSelect"
          role="button"
          tabindex="0"
        >
          <em>
            All car colors
          </em>
        </div>
        <input
          type="hidden"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiFormControl-root makeStyles-formControl-2"
      data-testid="NavFilterColorManufacturer"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink"
        data-shrink="true"
        id="NavFilterManufacturerLabel"
      >
        Manufacturer
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-selectEmpty-3 MuiInputBase-formControl MuiInput-formControl"
      >
        <div
          aria-haspopup="listbox"
          aria-labelledby="NavFilterManufacturerLabel NavFilterManufacturerLabelSelect"
          class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
          id="NavFilterManufacturerLabelSelect"
          role="button"
          tabindex="0"
        >
          <em>
            All manufacturers
          </em>
        </div>
        <input
          type="hidden"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-15 makeStyles-filterButtonContainer-4"
    >
      <a
        aria-disabled="false"
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge"
        data-testid="NavFilterButton"
        href="/?filters.color=&filters.manufacturer="
        role="button"
        tabindex="0"
      >
        <span
          class="MuiButton-label"
        >
          Filter
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </a>
    </div>
  </section>
</DocumentFragment>
`;
